BLD: Shave another 5-6 minutes off combined Travis jobs

moreati opened the pull request:
- Install wheel in the virtualenv, so pip builds Cython as a wheel & caches it
- Enable Cython compilation when the result will be cached
- Skip `pip install` outside chroot when we're building numpy in a chroot

The total time for all jobs drops from about 45 minutes to about [40 minutes](https://travis-ci.org/moreati/numpy/builds/74499272), once the Cython wheels have been cached.

rgommers commented:
Thanks @moreati. This seems to work, however I'm a bit puzzled by why skipping the Cython compile doesn't break anything and why the TravisCI tests report that the numpy version number for 26f7965 is ``numpy-1.11.0.dev0+33f4306``. Is it not rebuilding properly, or is it caching the generated ``version.py``?

njsmith commented:
"Skipping the cython compile" = the cython compiler is written in python, and by default it tries to compile *itself* to C when installed, so that it can later compile things faster. This takes ages, and is a huge waste of time for a CI system where we are only going to run the compiler a few times before throwing it away, so a 10% or whatever speedup does not justify spending multiple minutes building the thing.

I think the mismatch in revision ids is because Travis does not test the PR revision, but rather tests merge(PR, master).

moreati commented:
With this PR pip downloads/builds and *caches* all the requirements as a `.whl`. Critically (due to https://github.com/numpy/numpy/commit/b4ee7287c8a260c5ca4325b63325b6e66647c08b) Jenkins now saves this cache between job runs.

The cost of compiling Cython is only incurred once per new release of Cython, not every build. Normally pip will use the `.whl` it previously compiled. That's where some of the 10% speedup comes from.

Compiling Cython is a necessary condition to getting a `.whl`, pip doesn't create a wheel when build options (such as `--install-option=--no-cython-compile`) are present on the command line

homu commented:
:umbrella: The latest upstream changes (presumably #6429) made this pull request unmergeable. Please resolve the merge conflicts.

moreati commented:
Before I do so, does the project want this? The PR has been sitting untouched for over a month.

rgommers commented:
@moreati it looks like an oversight that we didn't merge this yet, sorry. If it's faster then we do want this. I'll merge this as soon as you've updated it.

rgommers commented:
(please comment when updated; Github doesn't notify on new commits)

moreati committed:
BLD: Speedup Cython install & run on Travis

This changeset
 - Installs wheel, so pip installs numpy dependencies as .whls - saving them
   to the Travis cache between builds.
 - Alows compilation of Cython, which allows pip to create a .whl of Cython
   that is similarly cached between runs
 - Disables compilation of Cython under CHROOT, since it runs with sudo=true
   and this feature disables the Travis caching

moreati commented:
I've rebased the patch, but the speedup has disappeared or gone slightly negative. I'm happy to close/abandon this PR.

rgommers commented:
Hmm, odd. Nothing changed in our ``.travis.yml`` file. You did trigger a second run so the cache was full and used right?

moreati commented:
> You did trigger a second run so the cache was full and used right?

Twice. [Before](https://travis-ci.org/moreati/numpy/builds/86236276) , [after #2](https://travis-ci.org/moreati/numpy/builds/86242118)

rgommers commented:
Then I guess we're closing this. Thanks for trying @moreati.

charris commented:
Might be related to the github switch to containers. Numpy support went in Aug 5.

moreati commented:
@charris do you mean the Numpy switch to containers on *Travis*? If so, not. It was me that wrote that and it was before I wrote this
